### PR TITLE
LCOE includes parasitic operating cost in numerator, energy sales in denominator

### LINKED ISSUE #532: Negative LPPA and LCOE for storage projects
SAM 2020.11.29 uses net energy to grid to calculate the LCOE and LPPA. For stand alone batteries which charge from the grid, this can lead to negative LCOE and LPPA values. Determine what kWh values should be used for these calculations in cases where the net energy is negative.

### PR DESCRIPTION
**Pull request by mjprilliman**
-Option to calculate parasitic operating cost from PPA price or retail rate for all FOM technologies (can be revisited for certain technologies)
-Warnings about using PPA price to calculate operating costs when IRR Target solution mode is selected
-Defaults updates (include utility rate variables for FOM financial models)
-Closes #532

### CODE REVIEW

**Comment by cpaulgilman** [comment]
The Cash Flow tab seems to be missing from the Results page in this branch. For example:

image

image

**Comment by mjprilliman** [comment]
> The Cash Flow tab seems to be missing from the Results page in this branch. For example:
> 
> image
> 
> image

There was a merge conflict in cashflow.lk that was preventing the script from compiling. Should be fixed now.

**Comment by cpaulgilman** [comment]
All of the PPA single owner configurations except PV Battery and Generic Battery have `ppa_soln_mode = 0` (Specify IRR target) as default. For the changes in this PR, I think we will need to change the default for all PPA configurations to `ppa_soln_mode = 1` (Specify PPA price), which means that we have to decide on an appropriate default PPA price for these configurations. One option is to use the PPA price result for the current defaults. Our decision on this should be documented in SAM-documentation to ensure it is preserved when we update defaults for the release: https://github.com/NREL/SAM-documentation/tree/master/Defaults%20References%20and%20Tools/Fall%202021

When the Electricity Purchases option is "Use PPA or market prices" (`ui_electricity_rate_option = 0`) with the Specify IRR target option (`ppa_soln_mode = 0`), the simulation runs without indicating that there is a problem in spite of the red text on the Electricity Purchases page: "The Use PPA or Market Prices option is not available with the Specify IRR Target option on the Revenue (or Financial Parameters) page. Either change the option to Specify PPA Price, or change the option here to Use Retail Electricity Rate(s)." If this is not a valid combination of inputs, I think SSC should generate an error.

I'm not sure how to validate the effect of this on LCOE, but I am seeing reasonable changes in LCOE, NPV, and cash flow outputs as I try different combinations of options, so I think this is OK to merge after the issues above are resolved.

File: deploy/runtime/startup.lk
Comment (brtietz) [inline deploy/runtime/startup.lk:746]: Typo here: utility rate 5 twice.

**Comment by mjprilliman** [comment]
> All of the PPA single owner configurations except PV Battery and Generic Battery have `ppa_soln_mode = 0` (Specify IRR target) as default. For the changes in this PR, I think we will need to change the default for all PPA configurations to `ppa_soln_mode = 1` (Specify PPA price), which means that we have to decide on an appropriate default PPA price for these configurations. One option is to use the PPA price result for the current defaults. Our decision on this should be documented in SAM-documentation to ensure it is preserved when we update defaults for the release: https://github.com/NREL/SAM-documentation/tree/master/Defaults%20References%20and%20Tools/Fall%202021
> 
> When the Electricity Purchases option is "Use PPA or market prices" (`ui_electricity_rate_option = 0`) with the Specify IRR target option (`ppa_soln_mode = 0`), the simulation runs without indicating that there is a problem in spite of the red text on the Electricity Purchases page: "The Use PPA or Market Prices option is not available with the Specify IRR Target option on the Revenue (or Financial Parameters) page. Either change the option to Specify PPA Price, or change the option here to Use Retail Electricity Rate(s)." If this is not a valid combination of inputs, I think SSC should generate an error.
> 
> I'm not sure how to validate the effect of this on LCOE, but I am seeing reasonable changes in LCOE, NPV, and cash flow outputs as I try different combinations of options, so I think this is OK to merge after the issues above are resolved.

@cpaulgilman  I introduced ssc errors for that combination in the FOM configs, and changed the default to the Specify PPA price option for all the configs although I still need to put in reasonable PPA prices like you were suggesting. The defaults are still a mess until I get some time to run the test script so I would wait on a re-review until that is resolved.

**Comment by brtietz** [comment]
Looks good overall, one nitpick:

On the cash flow tab "Net energy to grid" is no longer used in the LCOE calculations, this should be replaced with energy sales:

image

**Comment by mjprilliman** [comment]
> Looks good overall, one nitpick:
> 
> On the cash flow tab "Net energy to grid" is no longer used in the LCOE calculations, this should be replaced with energy sales:
> 
> image

Can somebody push this change? My build is tied up with the test script on another branch. It should just be changing cf_energy_net to cf_energy_sales on line 676 of cashflow.lk

image

**Comment by mjprilliman** [comment]
@brtietz @cpaulgilman I'm planning to merge this before there are more conflicts unless there are any objections

**Review by brtietz — APPROVED** [review]

**Pull request merged**
